Ensure thread id is valid in nested parallel regions

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60185 Ensure num_threads is initialized before calling omp_get_max_threads
* #60184 OpenMP: Refactor parallel_reduce to share code with parallel_for
* **#60183 Ensure thread id is valid in nested parallel regions**

Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

Differential Revision: [D29287816](https://our.internmc.facebook.com/intern/diff/D29287816)

peterbell10 committed:
Ensure thread id is valid in nested parallel regions

Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

ngimel commented:
Does only OpenMP need changes here? What about other parallel backends?
Also, TLS accesses are expensive, and while it looks that one is unavoidable here, it also means that unfortunately for perf reasons people will keep writing at the call site
```
if (there_is_not_a_lot_of_work){
f()
} else {
at::parallel_for(...)
}
```
No changes needed, but it's a pattern that you've eliminated recently in reduction, so now it became slightly more expensive.

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

peterbell10 commented:
> Does only OpenMP need changes here? What about other parallel backends?

I think your right, I've made similar changes to the other backends as well.

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

[ghstack-poisoned]

ngimel commented:
Great, can you please add a test for it (probably cpp test)?

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29287816).

peterbell10 committed:
Update on "Ensure thread id is valid in nested parallel regions"


Fixes https://github.com/pytorch/pytorch/pull/59149#issuecomment-863287331

`parallel_for` will call the function directly if it would have run on only a
single thread anyway. This is great for performance, but causes an issue in
nested parallel regions because `get_thread_num` will reflect the parent
parallel region instead of the current `parallel_for` call.

I fix this by using a `thread_local` variable for the current thread id and
manually setting it before each call to the user-provided function.

Differential Revision: [D29287816](https://our.internmc.facebook.com/intern/diff/D29287816)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/test/test_parallel.cpp:
```diff
@@ -9,12 +9,24 @@
 TEST(TestParallel, TestParallel) {
   manual_seed(123);
-  set_num_threads(1);
+  NumThreadsGuard guard(1);
```
This test was having the global side-effect of disabling multi-threading.

ngimel replied:
Thanks for fixing this!

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29287816).
